### DISCUSSION TITLE
How to view automatically generated CREATE TYPE statements for Enum types before executing create_all()?

### DISCUSSION THREAD

**Discussion by krave1986** [sqlalchemy/sqlalchemy#12166] [Usage Questions] [answered]
Hi team,

According to the official document [Using Python Enum or pep-586 Literal types in the type map](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map), when using SQLAlchemy ORM with PostgreSQL, `Enum(Status)` datatypes will be automatically linked with Python built-in `enum.Enum` types.

I followed the documentation and tried to check the output of CREATE TABLE. However, I noticed that the CREATE TYPE statement for the Enum type is missing from the result.

```
import enum

from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.schema import CreateTable
from sqlalchemy.dialects import postgresql


class Base(DeclarativeBase):
    pass


class Status(enum.Enum):
    PENDING = "pending"
    RECEIVED = "received"
    COMPLETED = "completed"


class SomeClass(Base):
    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[Status]


print(CreateTable(SomeClass.__table__).compile(dialect=postgresql.dialect()))
```

Expected output:

```
CREATE TYPE status AS ENUM ('PENDING', 'RECEIVED', 'COMPLETED')

CREATE TABLE some_table (
  id SERIAL NOT NULL,
  status status NOT NULL,
  PRIMARY KEY (id)
)
```

Actual output:

```
CREATE TABLE some_table (
        id INTEGER NOT NULL,        
        status VARCHAR(9) NOT NULL, 
        PRIMARY KEY (id)
)
```

I can confirm that `CREATE TYPE` statement automatically being created if I run this in echo mode:

```
Base.metadata.create_all(engine)
```

My question is: Is there a way to check to resulted SQL statements (including CREATE TYPE) before connecting PostgreSQL to ensure that Enum types created implicitly.

**Comment by CaselIT** [comment]
Hi,

That's expected. The `CreateTable` object is used to generate the `create table` statement only.
If you want to create a single table use the `Table.create()` method

**Reply by CaselIT** [reply]
As for your question, I don't think there is a public api that will tell give you all the statements that are run by the `create()` method.

You can of course use an event to see the statement before they are executed, but I'm not sure if that fits the bill.

**Comment by zzzeek** [comment, marked as answer]
you can run the whole metadata.create_all sequence into a string using [create_mock_engine](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_mock_engine)

```py
from sqlalchemy import create_mock_engine


def dump(sql, *multiparams, **params):
    print(sql.compile(dialect=engine.dialect))


engine = create_mock_engine("postgresql+psycopg2://", dump)
metadata.create_all(engine, checkfirst=False)
```

**Reply by CaselIT** [reply]
forgot it was a thing!
